### PR TITLE
Play order updates

### LINKED ISSUE #1111: Repeat-block based play mode modifier
Original [issue 1111](https://code.google.com/p/quodlibet/issues/detail?id=1111) created by _deepfire@feelingofgreen.ru on 2013-01-31T10:07:17.000Z:

1. Problem description

It is often the case, especially when listening to new music, that listening to a song once is not enough to get a satisfactory feel of it.

Yet, constant switching between non-progressing &quot;One song&quot; mode and progressing (either &quot;Shuffle&quot; or &quot;In order&quot;) modes is not ergonomic.

2. Solution

These conflicting needs could be balanced by introducing an integer &quot;granularity modifier&quot;, which would indicate how many times each song must be played, before next one is progressed into.

3. Implementation considerations

Note that this could be implemented as a straightforward generalisation of the current algorithm -- by simply stating, that the current behaviour has the &quot;granularity modifier&quot; set to 1.

What is important, is that this modifier is completely orthogonal to the play mode -- it is equally applicable to in-order, shuffle and weighted modes.  It does not have a semantic conflict with one-song mode either.

I have no idea what the &quot;follow cursor&quot; means, so cannot comment on that.

4. UI implications

The solution is envisioned as an integer spin-box, defaulted to 1, and presented near the &quot;Order&quot; and &quot;Repeat&quot; controls.

It could be named &quot;Song repeat&quot;.

This somewhat clashes with the &quot;Repeat&quot; checkbox, which could be renamed to &quot;Repeat playlist&quot;.

### LINKED ISSUE #1469: "Follow Cursor" plugin does not work
Original [issue 1469](https://code.google.com/p/quodlibet/issues/detail?id=1469) created by Ross.Swartz on 2014-09-23T01:45:20.000Z:

<b>Please describe the problem thoroughly and accurately. If possible paste</b>
<b>the contents of the ~/.quodlibet Dump or MiniDump file associated with the</b>
<b>problem.</b>

<b>- What did you try to do?</b>
Enable the follow cursor plugin

<b>- What did you expect to happen?</b>
The next song to play would be the song highlighted by the cursor (that isn't the currently playing song)

<b>- What did happen instead?</b>
Playback continued according to the playback mode selected with no effect

<b>- Which version of Quod Libet?</b>
3.2.1 (from apt-get)

<b>- Which operating system including version (Ubuntu 14.04, Win7, Debian</b>
<b>sid)?</b>
Ubuntu 14.04 with XFCE installed

### LINKED ISSUE #1836: Add Repeat-One feature
Currently only the whole playlist can be repeated. It’s common to be able to have `Repeat All` and `Repeat One`.

This could be done using three states: `No Repeat`, `Repeat One` and `Repeat All`, the `Repeat All` icon is mostly the repeat icon with a one-symbol, something like this:

Repeat-One-Icon

Repeat-Off could be the `Repeat All`-icon greyed out.

The inactive shuffle-icon should be greyed out then as well.

### PR DESCRIPTION
**Pull request by declension**
* Split `PlayOrderPlugin` to `ShufflePlugin` and `RepeatPlugin` marker subclasses
* Add `Repeat` as an explicit (delegating) `Order`, and `Reorder` as the base class for shuffle-style `Order`s (necessary for plugin manager). This then fixes #1111. 
* Add lightweight GObject-based list-like store for `Orders` that signals (but not quite `ListStore`), and a subclass that is also a `PluginManager`
* New generalised class `ToggledPlayOrderMenu`
* `PlayOrder` now `PlayOrderWidget` which tries to do a little less; it's about aggregating the repeat and shuffle controls, and exposing higher level information (notably, the single composed `Order`).
* Remove concept of repeat from Playlist, as repeating is more complex now and handled elsewhere.
* Add  "Repeat All" and "Repeat One" (fixes #1836) type repeaters.
* Config: Finally keep `memory.repeat` and `memory.shuffle` together (removing `settings.repeat`)
* Use more properties everywhere for readability
* Remove mixins that no longer help us
* Fix current plugins to work with new structure (Follow now working fine, fixes #1469).
* Add timeout to mpris test as it was hanging during dev

There _must_ be a few things I've missed / broken... but can't see them right now. 

### TODOs
 * [x] Single
 * [x] Fix / simplify `PlayerOptions`
 * [x] Remoting is broken a bit I think.

### CODE REVIEW

**Comment by lazka** [comment]
Just gave it a quick run:

* The check menu items in the tray icon aren't working
* The "follow cursor" and "queue only" only work when random is on. What's your plan there?

**Comment by declension** [comment]
Thanks! 



>   The check menu items in the tray icon aren't working
Cool I should look at that plugin, forgot how much it touches. 


>   The "follow cursor" and "queue only" only work when random is on. What's your plan there?

My plan was actually no plan really - to me they're shuffle modes, or "shufflers" (I considered renaming these _reorderers_ or _sorters_ but both terms aren't much better IMO), and as such these should be turned on and off by 

Though Queue Only is more interesting I guess.... (though I never use it)... it's harder to argue it's a shuffle mode. To me the _correct_ way here is abstracting the `PlaylistMux` layer a bit and providing the option here. Or - just having this as a top-level feature away from all these perhaps (easier in many ways)

Thoughts?

**Comment by lazka** [comment]
> Cool I should look at that plugin, forgot how much it touches. 

Afair all mpd/trayicon/mpris go through a simplified interface in the core.

> Thoughts?

Ah, I thought they work different now, never mind.

File: quodlibet/quodlibet/ext/events/mpris/mpris1.py [hunk 1]
```diff
@@ -90,6 +90,7 @@ def __init__(self):
         super(MPRIS1Player, self).__init__(name, self.PATH)
 
         player_options = app.player_options
+        # FIXME: this
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/mpris/mpris1.py]: hm?

File: quodlibet/quodlibet/ext/events/mpris/mpris2.py [hunk 2]
```diff
@@ -362,7 +362,7 @@ def get_property(self, interface, name):
             elif name == "Rate":
                 return 1.0
             elif name == "Shuffle":
-                return player_options.random
+                return player_options.shuffle
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/mpris/mpris2.py]: related: the mpd server is also affected by this rename and still broken

File: quodlibet/quodlibet/ext/events/randomalbum.py [hunk 3]
```diff
@@ -160,8 +160,8 @@ def _score(self, albums):
         return [(score, name) for name, score in scores.items()]
 
     def plugin_on_song_started(self, song):
-        if (song is None and config.get("memory", "order") != "onesong" and
-            not app.player.paused):
+        one_song = config.get("memory", "repeat") != "one_song"
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/randomalbum.py]: isn't memory/repeat a boolean?

File: quodlibet/quodlibet/qltk/playorder.py [hunk 4]
```diff
@@ -173,248 +230,345 @@ def next(self, playlist, iter):
...
+        toggle.show_all()
+        qltk.remove_padding(toggle)
+        toggle.set_size_request(26, 26)
+        self.pack_start(toggle, True, True, 0)
 
-    def set_shuffle(self, value):
-        """Set shuffle, will change the order accordingly"""
+        def forward_signal(*args):
+            if self.__inhibit:
+                print_d("NOT forwarding programmatic toggled signal...")
```
Comment (lazka) [inline quodlibet/quodlibet/qltk/playorder.py]: Did you want something like: https://lazka.github.io/pgi-docs/#GObject-2.0/functions.html#GObject.signal_handler_block ?

File: quodlibet/tests/test_qltk_songmodel.py [hunk 5]
```diff
@@ -144,35 +140,40 @@ def test_weighted(self):
...
-    def test_onesong_repeat(self):
-        self.pl.go_to(3)
-        self.pl.order = ORDERS[3](self.pl)
-        self.pl.repeat = True
-        self.failUnlessEqual(self.pl.current, 3)
-        self.pl.next()
-        self.failUnlessEqual(self.pl.current, 4)
-        self.pl.next_ended()
-        self.failUnlessEqual(self.pl.current, 4)
+    # def test_onesong(self):
```
Comment (lazka) [inline quodlibet/tests/test_qltk_songmodel.py]: I guess remove that?

File: quodlibet/tests/test_qltk_playorder.py [hunk 6]
```diff
@@ -3,72 +3,107 @@
...
+from quodlibet.qltk.playorder import PlayOrderWidget
+from tests import TestCase
 
 
-class TPlayOrder(TestCase):
+class TPlayOrderWidget(TestCase):
 
     def setUp(self):
-        quodlibet.plugins.init()
+        # quodlibet.plugins.init()
```
Comment (lazka) [inline quodlibet/tests/test_qltk_playorder.py]: please remove

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 7]
```diff
@@ -129,34 +122,36 @@ def single(self):
         When `repeat` is True the current song will be replayed.
         """
 
-        return self._order.get_active_name() == "onesong"
+        # FIXME: Add OneSong repeat mode etc
```
Comment (lazka) [inline quodlibet/quodlibet/qltk/quodlibetwindow.py]: I guess something like `repeat == True and repeate_mode == "repeat_song"`?

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 8]
```diff
@@ -129,34 +122,36 @@ def single(self):
...
+        return False
 
     @single.setter
     def single(self, value):
         if value and not self.single:
-            self._order.set_active_by_name("onesong")
```
Comment (lazka) [inline quodlibet/quodlibet/qltk/quodlibetwindow.py]: You can test through ``mpris-remote repeat true`` btw

or using mpdroid on android (in the top right corner menu you can enable single mode)

**Comment by lazka** [comment]
OK. mpd still crashes, but I can have a look myself later. lgtm

**Comment by declension** [comment]
Thanks @lazka, `master` is merged in and my MPD breakage should be fixed now (I installed `mpdroid` to test manually and it became clear). _Single_ is still a bit confusing but this is more because of the split of functionality.

Anyway assuming it goes green I'll merge later today unless you have any blockers.

**Comment by lazka** [comment]
thanks. sure, go ahead.

**Comment by lazka** [comment]
One more thing: the menus don't update when the class is changed from outside

**Pull request merged**